File: app/src/models/tip.ts
import { Branch } from './branch'
import { assertNever } from '../lib/fatal-error'

export enum TipState {
  Unknown = 'Unknown',
  Unborn = 'Unborn',
  Detached = 'Detached',
  Valid = 'Valid',
}

export interface IUnknownRepository {
  readonly kind: TipState.Unknown
}

export interface IUnbornRepository {
  readonly kind: TipState.Unborn

  /**
   * The symbolic reference that the unborn repository points to currently.
   *
   * Typically this will be "master" but a user can easily create orphaned
   * branches externally.
   */
  readonly ref: string
}

export interface IDetachedHead {
  readonly kind: TipState.Detached
  /**
   * The commit identifier of the current tip of the repository.
   */
  readonly currentSha: string
}

export interface IValidBranch {
  readonly kind: TipState.Valid
  /**
   * The branch information associated with the current tip of the repository.
   */
  readonly branch: Branch
}

export type Tip =
  | IUnknownRepository
  | IUnbornRepository
  | IDetachedHead
  | IValidBranch

/**
 * Gets a value indicating whether two Tip instances refer to the
 * same canonical Git state.
 */
export function tipEquals(x: Tip, y: Tip) {
  if (x === y) {
    return true
  }

  const kind = x.kind
  switch (x.kind) {
    case TipState.Unknown:
      return x.kind === y.kind
    case TipState.Unborn:
      return x.kind === y.kind && x.ref === y.ref
    case TipState.Detached:
      return x.kind === y.kind && x.currentSha === y.currentSha
    case TipState.Valid:
      return x.kind === y.kind && branchEquals(x.branch, y.branch)
    default:
      return assertNever(x, `Unknown tip state ${kind}`)
  }
}

function branchEquals(x: Branch, y: Branch) {
  return (
    x.type === y.type &&
    x.tip.sha === y.tip.sha &&
    x.remote === y.remote &&
    x.upstream === y.upstream
  )
}

File: GitDesktop/GitDesktop/Models/Tip.swift
import Foundation

/// Repository HEAD state. Port of `Tip` in `models/tip.ts`.
public enum Tip: Sendable, Equatable {
    case unknown
    case unborn(ref: String)
    case detached(currentSha: String)
    case valid(branch: Branch)

    public var kind: TipState {
        switch self {
        case .unknown: return .unknown
        case .unborn: return .unborn
        case .detached: return .detached
        case .valid: return .valid
        }
    }
}

public enum TipState: String, Codable, Sendable {
    case unknown = "Unknown"
    case unborn = "Unborn"
    case detached = "Detached"
    case valid = "Valid"
}

public func tipEquals(_ x: Tip, _ y: Tip) -> Bool {
    switch (x, y) {
    case (.unknown, .unknown): return true
    case (.unborn(let a), .unborn(let b)): return a == b
    case (.detached(let a), .detached(let b)): return a == b
    case (.valid(let a), .valid(let b)):
        return a.type == b.type
            && a.tip.sha == b.tip.sha
            && a.upstreamRemoteName == b.upstreamRemoteName
            && a.upstream == b.upstream
    default: return false
    }
}
